### PR TITLE
Extract a UILabel wrapper or subclass which displays Ethereum address and/or ENS name #1950

### LINKED ISSUE #1213: ENS loading issues
1. If the network is slow, users may submit a transition before ENS load up. It will end up with Invalid Address error message.

- [ ] need an Indication to tell users, the address is loading up.

2. Once triggered Invalid Address error message, the wallet will stop loading the ENS. Means once users click the send before ENS load up, the ENS will never load up again, unless you kill the app and restart it.

- [ ] need to continue loading.

### LINKED ISSUE #1950: Extract a UILabel wrapper or subclass which displays Ethereum address and/or ENS name
There at at least a few possibilities:

* Just address (sometimes the full address, sometimes shortened to 0xabcd...abcd)
* Just ENS
* "<ENS> | <address>" (in Settings screen)
* Maybe more?

Can you replace all of them with a single class with a way to configure them?

If it makes sense, let it be able to resolve, reverse resolve ENS names too, but might not be.

### LINKED ISSUE #1977: Pasting an address in Import Wallet should not wait until ENS resolution is complete before enabling the watch button
Also see if this affects other places where `AddressTextField` is used. Eg send screen.

### PR DESCRIPTION
**Pull request by vladyslav-iosdev**
Closes #1950, closes #1213, closes #1977

### CODE REVIEW

File: AlphaWallet/UI/AddressTextField.swift [hunk 1]
```diff
@@ -15,27 +15,16 @@ class AddressTextField: UIControl {
...
-        switch addressOrEnsName {
-        case .address(let value):
-            return value.eip55String
-        case .ensName, .none:
-            return nil
-        }
-    }
+    private let ensAddressLabel: AddressOrEnsNameLabel = {
+        let label = AddressOrEnsNameLabel()
+        label.addressFormat = .full
```
Comment (hboon) [inline AlphaWallet/UI/AddressTextField.swift:20]: This should be `.truncateMiddle`

**Comment by hboon** [comment]
Does this address both points in #1213 ? (if so you can mention "Closes #1213" too in the 1st comment in this PR).

**Comment by vladyslav-iosdev** [comment]
> Closes #1213"

yes i does

**Comment by hboon** [comment]
> > Closes #1213"
> 
> yes i does

You'll have to repeat the closes (or fixes keywords) like this: "Closes #xyz, closes #abc"

**Comment by vladyslav-iosdev** [comment]
> 1. hweeboon.eth

@hboon as i undarstand the problem is in that we can't cancel sent request. What if i wrap body of getENSAddressFromResolver into DispatchWorkItem that allow us to cancel previous sent getENSAddressFromResolver operation?

**Comment by hboon** [comment]
What if when we get the resolved result, we compare the "source" we have just resolved with the textfield contents? e.g (this is code from before the PR, I just added a line):

```
ENSReverseLookupCoordinator(server: serverToResolveEns).getENSNameFromResolver(forAddress: address) { [weak self] result in
    guard let strongSelf = self else { return }
    guard let resolvedESNname = result.value else {
        **COMPARE HERE `address` vs current value of textField **
        //Don't show an error when pasting what seems like a wrong ENS name for better usability
        strongSelf.addressOrEnsName = nil
        strongSelf.delegate?.didPaste(in: strongSelf)
        return
    }

    strongSelf.addressOrEnsName = .ensName(resolvedESNname)

    strongSelf.delegate?.didPaste(in: strongSelf)
}
```

Is this simpler?

**Comment by vladyslav-iosdev** [comment]
> What if when we get the resolved result, we compare the "source" we have just resolved with the textfield contents? e.g (this is code from before the PR, I just added a line):
> 
> ```
> ENSReverseLookupCoordinator(server: serverToResolveEns).getENSNameFromResolver(forAddress: address) { [weak self] result in
>     guard let strongSelf = self else { return }
>     guard let resolvedESNname = result.value else {
>         **COMPARE HERE `address` vs current value of textField **
>         //Don't show an error when pasting what seems like a wrong ENS name for better usability
>         strongSelf.addressOrEnsName = nil
>         strongSelf.delegate?.didPaste(in: strongSelf)
>         return
>     }
> 
>     strongSelf.addressOrEnsName = .ensName(resolvedESNname)
> 
>     strongSelf.delegate?.didPaste(in: strongSelf)
> }
> ```
> 
> Is this simpler?

yep, you right:) Done.

File: AlphaWallet/UI/AddressTextField.swift [hunk 2]
```diff
@@ -264,116 +228,43 @@ class AddressTextField: UIControl {
...
-                return
-            }
-
-            ENSReverseLookupCoordinator(server: serverToResolveEns).getENSNameFromResolver(forAddress: address) { [weak self] result in
-                guard let strongSelf = self else { return }
-                guard let resolvedESNname = result.value else {
-                    //Don't show an error when pasting what seems like a wrong ENS name for better usability
-                    strongSelf.addressOrEnsName = nil
-                    strongSelf.delegate?.didPaste(in: strongSelf)
+            ensAddressLabel.resolve(value) { [weak self] resolvation in
```
Comment (hboon) [inline AlphaWallet/UI/AddressTextField.swift:239]: "resolvation" -> "resolution"

File: AlphaWallet/Core/Views/AddressOrEnsNameLabel.swift [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    func resolve(_ value: String, completion: @escaping ((AddressOrEnsResolution) -> Void)) {
+        clear()
+        
+        if let address = AlphaWallet.Address(string: value) {
+            inResolvingState = true
+            ENSReverseLookupCoordinator(server: serverToResolveEns).getENSNameFromResolver(forAddress: address) { [weak self] result in
+                guard let strongSelf = self else { return }
+                strongSelf.inResolvingState = false
+
+                if let resolvedESNname = result.value {
```
Comment (hboon) [inline AlphaWallet/Core/Views/AddressOrEnsNameLabel.swift:113]: "resolvedESNname" -> "resolvedEnsName"

File: AlphaWallet/UI/AddressTextField.swift [hunk 4]
```diff
@@ -405,27 +296,39 @@ extension AddressTextField: UITextFieldDelegate {
...
+            self.ensAddressResolve(value: newValue)
         }
 
-        informDelegateDidChange(to: newValue ?? "")
+        informDelegateDidChange(to: newValue)
         return true
     }
 
+    private func ensAddressResolve(value: String) {
+        ensAddressLabel.resolve(value) { [weak self] resolvation in
```
Comment (hboon) [inline AlphaWallet/UI/AddressTextField.swift:313]: "resolvation -> "resolution" (but result might be more common?)

File: AlphaWallet/UI/AddressTextField.swift [hunk 5]
```diff
@@ -405,27 +296,39 @@ extension AddressTextField: UITextFieldDelegate {
...
+            guard let strongSelf = self, value == strongSelf.textField.text?.trimmed else {
+                return
+            }
+
+            switch resolvation {
```
Comment (hboon) [inline AlphaWallet/UI/AddressTextField.swift:318]: Seems like this switch can be extracted into a function. It's quite similar in `pasteAction()`, except for an error being displayed?

**Pull request merged**